packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts

```diff
@@ -575,6 +589,84 @@ export class PersistentCheckpointStateCache implements CheckpointStateCache {
     return Array.from(this.cache.keys());
   }
 
+  /**
+   * Prune or persist checkpoint states in an epoch, see the description in `processState()` function
+   */
+  private async processPastEpoch(
+    blockRootHex: RootHex,
+    state: CachedBeaconStateAllForks,
+    epoch: Epoch
+  ): Promise<number> {
+    let persistCount = 0;
+    const epochBoundarySlot = computeStartSlotAtEpoch(epoch);
+    const epochBoundaryRoot =
+      epochBoundarySlot === state.slot ? fromHexString(blockRootHex) : getBlockRootAtSlot(state, epochBoundarySlot);
+    const epochBoundaryHex = toHexString(epochBoundaryRoot);
+
+    // for each epoch, usually there are 2 rootHex respective to the 2 checkpoint states: Previous Root Checkpoint State and Current Root Checkpoint State
+    for (const rootHex of this.epochIndex.get(epoch) ?? []) {
+      const cpKey = toCacheKey({epoch: epoch, rootHex});
+      const cacheItem = this.cache.get(cpKey);
+
+      if (cacheItem !== undefined && isInMemoryCacheItem(cacheItem)) {
+        // this is state in memory, we don't care if the checkpoint state is already persisted
+        let {persistedKey} = cacheItem;
+        const {state} = cacheItem;
+        const logMeta = {
+          stateSlot: state.slot,
+          rootHex,
+          epochBoundaryHex,
+          persistedKey: persistedKey ? toHexString(persistedKey) : "",
+        };
+
+        if (rootHex === epochBoundaryHex) {
+          if (persistedKey) {
+            // no need to persist
+            this.logger.verbose("Pruned checkpoint state from memory but no need to persist", logMeta);
+          } else {
+            // persist and do not update epochIndex
+            this.metrics?.statePersistSecFromSlot.observe(this.clock?.secFromSlot(this.clock?.currentSlot ?? 0) ?? 0);
+            const cpPersist = {epoch: epoch, root: epochBoundaryRoot};
+            let bufferPoolKey: number | undefined = undefined;
+            try {
+              const timer = this.metrics?.statePersistDuration.startTimer();
+              const stateBytesWithKey = this.serializeState(state);
+              bufferPoolKey = stateBytesWithKey.key;
+              persistedKey = await this.datastore.write(cpPersist, stateBytesWithKey.data);
+              timer?.();
+            } finally {
+              if (bufferPoolKey !== undefined) {
+                this.bufferPool?.free(bufferPoolKey);
+              }
+            }
```

Comment:
this pattern seems like the ideal usecase for the `using` pattern from
- https://github.com/tc39/proposal-explicit-resource-management
- https://medium.com/@bagherani/ecmascript-explicit-resource-management-early-implementation-in-typescript-5-2-5e4d08b2aee3
```suggestion
            {
              const timer = this.metrics?.statePersistDuration.startTimer();
              using stateBytesWithKey = this.serializeState(state);
              bufferPoolKey = stateBytesWithKey.key;
              persistedKey = await this.datastore.write(cpPersist, stateBytesWithKey.data);
              timer?.();
            }
```

I'm not sure that we should use this feature, but it's worth knowing about.

Reply:
@wemeetagain thanks it's a nice feature and really suitable for this code place, implemented it 👍